lib/model/devicedownloadstate.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (C) 2015 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package model
+
+import (
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+type deviceFolderFileDownloadState struct {
+	blockIndexes []int32
+	version      protocol.Vector
+}
+
+type deviceFolderDownloadState struct {
+	mut   sync.RWMutex
+	files map[string]*deviceFolderFileDownloadState
```

Comment:
I think this would be better as a `map[string]deviceFolderFileDownloadState` (that type sure is a mouthful). It makes it a little bit more annoying to update the map (need to assign the changed object back into the map instead of just accessing it via the pointer) but it saves maybe a whole bunch of pointers and makes the GC a bit happier.